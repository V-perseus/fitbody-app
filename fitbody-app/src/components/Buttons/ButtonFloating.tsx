import React from 'react'
import { Pressable, Text, TextStyle, View, ViewStyle } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'

import globals from '../../config/globals'

interface IButtonFloatingProps {
  style?: ViewStyle
  textStyle?: TextStyle
  onPress: () => void
  text: string
  gradientColors?: string[]
  pgColor?: string
}
export const ButtonFloating: React.FC<IButtonFloatingProps> = ({ style, textStyle, onPress, text, gradientColors, pgColor, ...rest }) => {
  const shadowColor = pgColor || gradientColors?.[0] || globals.styles.colors.colorGrayDark

  return (
    <Pressable
      style={({ pressed }) => [styles.buttonStyle(pressed, pgColor), { shadowColor }, style]}
      onPress={onPress}
      {...rest}
      testID="button_floating">
      {gradientColors && gradientColors?.length > 1 ? (
        <LinearGradient colors={gradientColors} style={styles.innerButton}>
          <Text style={[styles.textStyle, textStyle]}>{text}</Text>
        </LinearGradient>
      ) : (
        <View style={styles.innerButton}>
          <Text style={[styles.textStyle, textStyle]}>{text}</Text>
        </View>
      )}
    </Pressable>
  )
}

const styles = {
  buttonStyle: (pressed: boolean, pgColor: string | undefined): ViewStyle => ({
    flexDirection: 'row',
    height: 56,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 28,
    marginHorizontal: 24,
    shadowOpacity: pressed ? 0.3 : 0.7,
    shadowOffset: { width: 0, height: 3 },
    shadowRadius: pressed ? (pgColor ? 5 : 5) : pgColor ? 5 : 15,
    elevation: pressed ? 4 : 13,
    backgroundColor: globals.styles.colors.colorPink,
  }),
  innerButton: {
    flex: 1,
    borderRadius: 28,
    height: 56,
    alignItems: 'center',
    justifyContent: 'center',
  } as ViewStyle,
  textStyle: {
    color: globals.styles.colors.colorWhite,
    fontSize: 16,
    ...globals.styles.fonts.primary.semiboldStyle,
  } as ViewStyle,
}
