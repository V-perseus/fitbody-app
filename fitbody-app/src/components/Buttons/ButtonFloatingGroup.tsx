import React from 'react'
import { Text, View, Pressable, ViewStyle } from 'react-native'
import { scale, moderateScale } from 'react-native-size-matters'

import BottomHover from '../../shared/BottomHover'

import ShareIcon from '../../../assets/images/svg/icon/24px/share.svg'
import EditIcon from '../../../assets/images/svg/icon/24px/edit.svg'
import CopyIcon from '../../../assets/images/svg/icon/24px/copy.svg'
import TrashIcon from '../../../assets/images/svg/icon/24px/trash.svg'

import globals from '../../config/globals'

const AvailableIconNames = Object.freeze({
  share: ShareIcon,
  edit: EditIcon,
  copy: CopyIcon,
  trash: TrashIcon,
})
interface IButtonFloatingProps {
  onPressLeft: null | (() => void)
  onPressRight: null | (() => void)
  btnLeftStyles?: ViewStyle
  btnRightStyles?: ViewStyle
  btnLeftIconName?: keyof typeof AvailableIconNames
  btnRightIconName?: keyof typeof AvailableIconNames
  btnLeftText: string
  btnRightText: string
}
export const ButtonFloatingGroup: React.FC<IButtonFloatingProps> = ({
  onPressLeft,
  onPressRight,
  btnLeftStyles,
  btnRightStyles,
  btnLeftIconName = 'share',
  btnRightIconName = 'edit',
  btnLeftText = 'LEFT',
  btnRightText = 'RIGHT',
}) => {
  const leftButtonStyles = (pressed: boolean) => [styles.touchableLeft(pressed), btnLeftStyles && btnLeftStyles]
  const rightButtonStyles = (pressed: boolean) => [styles.touchableRight(pressed), btnRightStyles && btnRightStyles]
  const LeftIcon = AvailableIconNames[btnLeftIconName]
  const RightIcon = AvailableIconNames[btnRightIconName]

  return (
    <BottomHover>
      <View style={styles.container}>
        <Pressable onPress={onPressLeft} style={({ pressed }) => leftButtonStyles(pressed)} testID="button_left">
          <LeftIcon width={scale(24)} height={scale(24)} color={styles.iconColor} />
          <Text style={styles.touchableLeftText}>{btnLeftText}</Text>
        </Pressable>
        <Pressable onPress={onPressRight} style={({ pressed }) => rightButtonStyles(pressed)} testID="button_right">
          <RightIcon width={scale(24)} height={scale(24)} color={styles.iconColor} />
          <Text style={styles.touchableRightText}>{btnRightText}</Text>
        </Pressable>
      </View>
    </BottomHover>
  )
}

const styles = {
  container: { flexDirection: 'row', paddingHorizontal: 16 } as ViewStyle,
  touchableLeft: (pressed: boolean): ViewStyle => ({
    flex: 1,
    flexDirection: 'row',
    marginVertical: scale(5),
    marginHorizontal: scale(8),
    marginTop: scale(25),
    marginBottom: scale(30),
    height: scale(56),
    borderRadius: scale(28),
    backgroundColor: globals.styles.colors.colorSkyBlue,
    justifyContent: 'center',
    alignItems: 'center',
    shadowColor: globals.styles.colors.colorSkyBlue,
    shadowOpacity: pressed ? 0.3 : 0.7,
    shadowRadius: pressed ? 5 : scale(15),
    shadowOffset: { width: 0, height: 0 },
    elevation: 3,
  }),
  touchableLeftText: {
    marginLeft: scale(8),
    fontFamily: globals.fonts.primary.semiboldStyle.fontFamily,
    fontSize: moderateScale(16),
    textAlign: 'center',
    color: globals.styles.colors.colorWhite,
  } as ViewStyle,
  touchableRight: (pressed: boolean): ViewStyle => ({
    flex: 1,
    marginVertical: scale(5),
    marginHorizontal: scale(8),
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: scale(25),
    marginBottom: scale(30),
    height: scale(56),
    borderRadius: scale(28),
    backgroundColor: globals.styles.colors.colorLove,
    justifyContent: 'center',
    shadowColor: globals.styles.colors.colorPink,
    shadowOffset: { width: 0, height: 0 },
    shadowOpacity: pressed ? 0.3 : 0.7,
    shadowRadius: pressed ? 5 : scale(15),
    elevation: 3,
  }),
  touchableRightText: {
    marginLeft: scale(8),
    fontFamily: globals.fonts.primary.semiboldStyle.fontFamily,
    fontSize: moderateScale(16),
    textAlign: 'center',
    color: globals.styles.colors.colorWhite,
  } as ViewStyle,
  iconColor: globals.styles.colors.colorWhite,
}
