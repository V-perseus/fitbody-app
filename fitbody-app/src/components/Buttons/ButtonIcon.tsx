import React from 'react'
import { Pressable, Text, StyleSheet, ViewStyle, TextStyle } from 'react-native'
import globals from '../../config/globals'

interface IButtonIconProps {
  style?: ViewStyle
  textStyle?: TextStyle
  onPress: (() => void) | null
  text: string
  useOpacity?: boolean
  pressedOpacity?: number
  rightIcon?: () => JSX.Element
  leftIcon?: () => JSX.Element
}
export const ButtonIcon: React.FC<IButtonIconProps> = ({
  style,
  textStyle,
  onPress,
  text,
  useOpacity = true,
  pressedOpacity = 0.7,
  rightIcon = null,
  leftIcon = null,
  ...rest
}) => {
  return (
    <Pressable
      style={({ pressed }) => [{ opacity: pressed && useOpacity ? pressedOpacity : 1 }, styles.buttonStyle, style]}
      onPress={onPress}
      {...rest}>
      {leftIcon && leftIcon()}
      <Text style={[styles.textStyle, textStyle]}>{text}</Text>
      {rightIcon && rightIcon()}
    </Pressable>
  )
}

const styles = StyleSheet.create({
  buttonStyle: {
    width: 114,
    height: 56,
    borderRadius: 8,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textStyle: {
    fontSize: 16,
    color: globals.styles.colors.colorWhite,
  },
})
