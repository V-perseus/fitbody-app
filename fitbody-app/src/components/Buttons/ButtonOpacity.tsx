import React from 'react'
import { Pressable, StyleSheet, ViewStyle } from 'react-native'

interface IButtonOpacityProps {
  style?: ViewStyle
  children: React.ReactNode
  onPress: () => void
  useOpacity?: boolean
  pressedOpacity?: number
  testID?: string
}
export const ButtonOpacity: React.FC<IButtonOpacityProps> = ({
  style,
  children,
  onPress,
  useOpacity = true,
  pressedOpacity = 0.5,
  ...rest
}) => {
  return (
    <Pressable
      style={({ pressed }) => [{ opacity: pressed && useOpacity ? pressedOpacity : 1 }, styles.buttonStyle, style]}
      onPress={onPress}
      hitSlop={16}
      {...rest}>
      {children}
    </Pressable>
  )
}

const styles = StyleSheet.create({
  buttonStyle: {
    alignItems: 'center',
    justifyContent: 'center',
  },
})
