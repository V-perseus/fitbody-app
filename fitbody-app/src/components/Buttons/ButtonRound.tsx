import React from 'react'
import { Pressable, ViewStyle, Text, StyleSheet, TextStyle } from 'react-native'
import globals from '../../config/globals'

interface IButtonRoundProps {
  style?: ViewStyle
  textStyle?: TextStyle
  onPress: () => void
  text: string
  icon?: React.ReactNode
  useOpacity?: boolean
  pressedOpacity?: number
}
export const ButtonRound: React.FC<IButtonRoundProps> = ({
  style,
  textStyle,
  onPress,
  text,
  icon,
  useOpacity = true,
  pressedOpacity = 0.7,
  ...rest
}) => {
  return (
    <Pressable
      style={({ pressed }) => [{ opacity: pressed && useOpacity ? pressedOpacity : 1 }, styles.buttonStyle, style]}
      onPress={onPress}
      {...rest}>
      {icon}
      <Text style={[styles.textStyle, textStyle]}>{text}</Text>
    </Pressable>
  )
}

const styles = StyleSheet.create({
  buttonStyle: {
    flexDirection: 'row',
    items: 'center',
    width: 114,
    height: 56,
    borderRadius: 28,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textStyle: {
    fontSize: 16,
    color: globals.styles.colors.colorWhite,
  },
})
