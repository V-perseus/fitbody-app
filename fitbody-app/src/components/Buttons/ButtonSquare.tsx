import React from 'react'
import { Pressable, Text, StyleSheet, ViewStyle, TextStyle, ActivityIndicator } from 'react-native'
import globals from '../../config/globals'

interface IButtonSquareProps {
  style?: ViewStyle
  textStyle?: TextStyle
  onPress: () => void
  text: string
  useOpacity?: boolean
  pressedOpacity?: number
  loading?: boolean
  testID?: string
}
export const ButtonSquare: React.FC<IButtonSquareProps> = ({
  style,
  textStyle,
  onPress,
  text,
  useOpacity = true,
  pressedOpacity = 0.7,
  loading = false,
  ...rest
}) => {
  return (
    <Pressable
      style={({ pressed }) => [{ opacity: pressed && useOpacity ? pressedOpacity : 1 }, styles.buttonStyle, style]}
      onPress={onPress}
      {...rest}>
      {loading ? (
        <ActivityIndicator size="small" color={globals.styles.colors.colorWhite} />
      ) : (
        <Text style={[styles.textStyle, textStyle]}>{text}</Text>
      )}
    </Pressable>
  )
}

const styles = StyleSheet.create({
  buttonStyle: {
    width: 114,
    height: 56,
    borderRadius: 8,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textStyle: {
    fontSize: 16,
    color: globals.styles.colors.colorWhite,
  },
})
