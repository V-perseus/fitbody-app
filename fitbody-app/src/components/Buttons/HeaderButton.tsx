import React from 'react'
import { Pressable, StyleProp, StyleSheet, ViewStyle } from 'react-native'

import BackArrowIcon from '../../../assets/images/svg/icon/24px/arrow-left.svg'
import globals from '../../config/globals'

interface IHeaderButtonProps {
  style?: StyleProp<ViewStyle>
  iconColor?: string
  onPress: () => void
  pressedOpacity?: number
  pressColor?: string
  children?: React.ReactNode
}
export const HeaderButton = ({
  style,
  iconColor = globals.styles.colors.colorBlack,
  onPress,
  pressedOpacity = 0.7,
  pressColor = 'rgba(0, 0, 0, 0.05)',
  children,
  ...rest
}: IHeaderButtonProps) => {
  return (
    <Pressable
      style={({ pressed }) => [
        { backgroundColor: pressed ? pressColor : 'transparent', opacity: pressed ? pressedOpacity : 1 },
        styles.buttonStyle,
        style,
      ]}
      hitSlop={16}
      onPress={onPress}
      {...rest}>
      {children ? children : <BackArrowIcon color={iconColor} testID="back_arrow"/>}
    </Pressable>
  )
}

const styles = StyleSheet.create({
  buttonStyle: {
    display: 'flex',
    width: 42,
    height: 42,
    borderRadius: 30,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: 7,
  },
})
