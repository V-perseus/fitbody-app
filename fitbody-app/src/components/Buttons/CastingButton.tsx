import React, { useEffect, useState } from 'react'
import { Pressable, ScrollView, StyleSheet, Text, View } from 'react-native'
import GoogleCast, { useDevices, useCastDevice, CastContext, CastButton } from 'react-native-google-cast'
import Modal from 'react-native-modal'

import CastIcon from '../../../assets/images/svg/icon/24px/cast/cast.svg'
import CastActiveIcon from '../../../assets/images/svg/icon/24px/cast/cast-active.svg'
import CastDevice from '../../../assets/images/svg/icon/24px/cast/cast-device.svg'

import globals from '../../config/globals'

import { ButtonOpacity } from './ButtonOpacity'

interface ICastingButtonProps {
  size?: number
  color?: string
}
const CastingButton: React.FC<ICastingButtonProps> = ({ size = 28, color = globals.styles.colors.colorWhite }) => {
  const castDevice = useCastDevice()
  const devices = useDevices()
  const sessionManager = GoogleCast.getSessionManager()
  const discoveryManager = GoogleCast.getDiscoveryManager()

  const [showDevicesModal, setShowDevicesModal] = useState(false)

  useEffect(() => {
    discoveryManager.startDiscovery()
    return () => {
      discoveryManager.stopDiscovery()
    }
  }, [discoveryManager])

  // console.log('DEVICES', devices)

  async function handleCast(deviceId: string) {
    try {
      const active = deviceId === castDevice?.deviceId
      if (active) {
        await sessionManager.endCurrentSession()
      } else {
        await sessionManager.startSession(deviceId)
      }

      setShowDevicesModal(false)
    } catch (error) {
      setShowDevicesModal(false)
    }
  }

  return (
    <>
      {/*
        Having the CastButton hidden here initializes a bunch of magic as far as permissions, session creation, and auto device discovery
        however it has issues responding to press events when a child of another pressable.
      */}
      <CastButton style={{ display: 'none', opacity: 0, width: 0, height: 0, tintColor: 'transparent' }} />
      <ButtonOpacity
        onPress={() => {
          // setShowDevicesModal(true)
          CastContext.showCastDialog()
        }}
        style={{ width: size, height: size }}>
        {devices.find((device) => device.deviceId === castDevice?.deviceId) ? (
          <CastActiveIcon color={color} style={{ width: size, height: size }} />
        ) : (
          <CastIcon color={color} style={{ width: size, height: size }} />
        )}
      </ButtonOpacity>
      <Modal
        onBackdropPress={() => setShowDevicesModal(false)}
        onBackButtonPress={() => setShowDevicesModal(false)}
        backdropOpacity={0.5}
        useNativeDriver={true}
        animationIn="fadeIn"
        animationOut="fadeOut"
        isVisible={showDevicesModal}>
        <View style={styles.containerView}>
          <Text style={styles.title}>Cast to</Text>
          <ScrollView style={{ width: '100%' }} contentContainerStyle={styles.deviceList}>
            {devices?.length > 0 ? (
              devices.map((device) => (
                <Pressable key={device.deviceId} onPress={() => handleCast(device.deviceId)}>
                  <View style={styles.deviceListItem}>
                    <CastDevice color={globals.styles.colors.colorBlack} style={{ marginRight: 12 }} />
                    <Text style={styles.deviceListItemText}>{device.friendlyName}</Text>
                  </View>
                </Pressable>
              ))
            ) : (
              <Text style={styles.deviceListItem}>No devices found</Text>
            )}
          </ScrollView>
        </View>
      </Modal>
    </>
  )
}

const styles = StyleSheet.create({
  containerView: {
    display: 'flex',
    paddingTop: 0,
    padding: 24,
    maxHeight: globals.window.width * 0.7,
    width: globals.window.width * 0.8,
    backgroundColor: globals.styles.colors.colorWhite,
    borderRadius: 8,
    alignSelf: 'center',
  },
  title: {
    marginVertical: 10,
    fontSize: 18,
    fontFamily: globals.fonts.primary.boldStyle.fontFamily,
    letterSpacing: -0.43,
  },
  deviceList: {
    flexGrow: 1,
    width: '100%',
    justifyContent: 'flex-start',
    paddingTop: 12,
  },
  deviceListItem: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 12,
  },
  deviceListItemText: {},
})

export default CastingButton
