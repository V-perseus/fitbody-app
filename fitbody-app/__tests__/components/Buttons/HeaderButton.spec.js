import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'
import renderer from 'react-test-renderer'

import { HeaderButton } from '../../../src/components/Buttons/HeaderButton'
import ChevronRight from '../../../assets/images/svg/icon/16px/cheveron/right.svg'

describe('HeaderButton', () => {
  const onPress = jest.fn()

  it('renders snapshot correctly', () => {
    const tree = renderer.create(<HeaderButton onPress={onPress} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render back arrow icon', () => {
    const { getByTestId } = render(<HeaderButton onPress={onPress} />)
    expect(getByTestId('back_arrow')).not.toBeNull()
  })

  it('should handle press event', async () => {
    const { getByTestId } = render(<HeaderButton onPress={onPress} testID="btn" />)
    await fireEvent.press(getByTestId('btn'))
    expect(onPress).toHaveBeenCalled()
  })

  it('should render custom icon', async () => {
    const { queryByTestId } = render(
      <HeaderButton onPress={onPress}>
        <ChevronRight testID="icon" />
      </HeaderButton>,
    )
    expect(queryByTestId('icon')).not.toBeNull()
  })
})
