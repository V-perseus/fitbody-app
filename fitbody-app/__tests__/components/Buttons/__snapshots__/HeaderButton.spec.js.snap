// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderButton renders snapshot correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={16}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "transparent",
        "opacity": 1,
      },
      {
        "alignItems": "center",
        "borderRadius": 30,
        "display": "flex",
        "height": 42,
        "justifyContent": "center",
        "marginHorizontal": 7,
        "width": 42,
      },
      undefined,
    ]
  }
>
  <arrow-left.svg
    color="#333333"
    testID="back_arrow"
  />
</View>
`;
